### PR TITLE
Campaign actions user input validation

### LINKED ISSUE #522: Request input binding and validation
We'll need to add user input validation on all of our action endpoints in a standardized way.

Use gin's model binding and validation as described here: https://github.com/gin-gonic/gin#model-binding-and-validation

### PR DESCRIPTION
**Pull request by Dzalevski**
# Pull Request Template

## Description

Rework Campaigns action to validate user input params.

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A - tested by insomnia calls

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: actions/campaigns.go [hunk 1]
```diff
@@ -24,8 +25,8 @@ import (
 
 type sendCampaignParams struct {
 	Ids      []int64 `form:"segment_id[]" valid:"required"`
-	Source   string  `form:"source" valid:"email,required~Email is empty or in invalid format"`
-	FromName string  `form:"from_name" valid:"required,stringlength(1|191)~From name is blank or exceeds maximum character limit."`
+	Source   string  `form:"source" binding:"required"`
```
Comment (FilipNikolovski) [inline actions/campaigns.go:28]: @Dzalevski the source should be a valid `email` address.

File: actions/campaigns.go [hunk 2]
```diff
@@ -222,11 +210,22 @@ func GetCampaign(c *gin.Context) {
 	})
 }
 
+type paramsCampaign struct {
```
Comment (FilipNikolovski) [inline actions/campaigns.go:213]: Rename paramsCampaign to postCampaign, to keep consistency.

File: actions/campaigns.go [hunk 3]
```diff
@@ -38,26 +39,13 @@ func StartCampaign(c *gin.Context) {
...
-		return
-	}
-
-	v, err := valid.ValidateStruct(params)
-	if !v {
-		msg := "Unable to start campaign, invalid request parameters."
-		if err != nil {
-			msg = err.Error()
+	if err := c.ShouldBind(params); err != nil {
+		for _, fieldErr := range err.(validator.ValidationErrors) {
```
Comment (FilipNikolovski) [inline actions/campaigns.go:43]: Replace this with AbortWithError().

**Review by FilipNikolovski — CHANGES_REQUESTED** [review]

File: actions/campaigns.go (hunk 2 above)
Comment (Dzalevski) [reply actions/campaigns.go:213]: @FilipNikolovski I'm using params coz it's same structure for put and post camp should i make 2 same structures?

File: validator/fielderror.go [hunk 4]
```diff
@@ -14,11 +14,19 @@ func (q FieldError) String() string {
 	var sb strings.Builder
 
 	sb.WriteString("Validation failed on field '" + q.Err.Field() + "'")
-	sb.WriteString(", condition: " + q.Err.ActualTag())
 
-	// Print condition parameters, e.g. max=191 -> max { 191 }
-	if q.Err.Param() != "" {
-		sb.WriteString(" " + q.Err.Param())
+	switch q.Err.ActualTag() {
```
Comment (Dzalevski) [inline validator/fielderror.go:18]: @FilipNikolovski  we have added custom err msgs for all tags can you take a look at the msgs? @Gligor23

File: actions/campaigns.go (hunk 2 above)
Comment (FilipNikolovski) [reply actions/campaigns.go:206]: One structure is okay in this case, we can rename it to `campaignParams` to be consistent.

File: validator/fielderror.go [hunk 5]
```diff
@@ -14,11 +14,19 @@ func (q FieldError) String() string {
...
+	case "email":
+		sb.WriteString(", wrong " + q.Err.ActualTag() + " format")
```
Comment (FilipNikolovski) [inline validator/fielderror.go:20]: Here we know that ActualTag() is "email", we don't need to concatenate strings here, just write the whole message: `", wrong email format"`. The same goes for the rest of the messages. Other than that, this is great 👍

**Review by FilipNikolovski — CHANGES_REQUESTED** [review]

File: validator/fielderror.go (hunk 5 above)
Comment (Dzalevski) [reply validator/fielderror.go:20]: @FilipNikolovski  this is only for 'email' and 'required' , for 'max' and 'min'  I think it's okay.

**Review by FilipNikolovski — APPROVED** [review]

**Pull request merged**